sdk/compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs

```diff
@@ -536,25 +540,84 @@ generatePackageMap version mbProjRoot userPkgDbs = do
                      (LF.extPackagePkg $ LF.dalfPackagePkg pkg)
                      (LF.dalfPackageId pkg)
 
+getUpgradedPackageErrs :: Options -> LSP.NormalizedFilePath -> LF.Package -> [FileDiagnostic]
+getUpgradedPackageErrs opts file mainPkg = catMaybes $
+  [ justIf (not $ optDamlLfVersion opts `LF.supports` LF.featurePackageUpgrades) $
+      ideErrorPretty file $ mconcat
+        [ "Main package LF Version ("
+        , T.pack $ LF.renderVersion $ optDamlLfVersion opts
+        , ") does not support Smart Contract Upgrades"
+        ]
+  , justIf (not $ LF.packageLfVersion mainPkg `LF.supports` LF.featurePackageUpgrades) $
+      ideErrorPretty file $ mconcat
+        [ "Upgraded package LF Version ("
+        , T.pack $ LF.renderVersion $ LF.packageLfVersion mainPkg
+        , ") does not support Smart Contract Upgrades"
+        ]
+  , justIf (optDamlLfVersion opts `lfVersionLt` LF.packageLfVersion mainPkg) $
+      ideErrorPretty file $ mconcat
+        [ "Upgraded package LF Version ("
+        , T.pack $ LF.renderVersion $ LF.packageLfVersion mainPkg
+        , ") cannot be higher than main package LF Version ("
+        , T.pack $ LF.renderVersion $ optDamlLfVersion opts
+        , ")"
+        ]
+  ] ++ case LF.packageMetadata mainPkg of
+    Nothing -> [Just $ ideErrorPretty file ("Upgraded DAR does not contain metadata" :: T.Text)]
+    Just meta ->
+      [ justIf (optMbPackageName opts /= Just (LF.packageName meta)) $
+          ideErrorPretty file $ T.unlines
+            [ "Upgraded package must have the same package name as main package."
+            , "Expected " <> maybe "<unknown>" LF.unPackageName (optMbPackageName opts)
+            , "Got " <> LF.unPackageName (LF.packageName meta)
+            ]
+      , justIf (optMbPackageVersion opts == Just (LF.packageVersion meta)) $
+          ideErrorPretty file $ mconcat
+            [ "Upgraded package cannot have the same package version as main package ("
+            , LF.unPackageVersion $ LF.packageVersion meta
+            , ")"
+            ]
+      , justIf (maybe False (`packageVersionLt` LF.packageVersion meta) $ optMbPackageVersion opts) $
+          ideErrorPretty file $ mconcat
+            [ "Upgraded package cannot have a higher package version ("
+            , LF.unPackageVersion $ LF.packageVersion meta
+            , ") than main package ("
+            , LF.unPackageVersion $ fromJust $ optMbPackageVersion opts
+            , ")"
+            ]
+      ]
+  where
+    justIf :: Bool -> a -> Maybe a
+    justIf cond val = guard cond >> Just val
+
+    packageVersionLt :: LF.PackageVersion -> LF.PackageVersion -> Bool
+    packageVersionLt = (<) `on` fmap (read @Int) . splitWhen (=='.') . T.unpack . LF.unPackageVersion
+
+    -- We assume major will always be 1 (or at least the same), since Version.supportedOutputVersions does not contain any 2. versions
+    -- TODO: Is this acceptable?
```

Comment:
Double check this

Reply:
I think it's still possible for a user to supply a 2.x DAR via the upgrades field and for it to be deserialized, but a different check ahead of time during deserialization may stop this code path from being reached.

I think it's a case worth handling with a specific error anyways, since it shouldn't take much time - just check beforehand that both DARs have the same major LF version and error out via justIf if not.